File: backend-app/src/main/java/com/gradientgeeks/ageis/backendapp/entity/Transaction.java
package com.gradientgeeks.ageis.backendapp.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Transaction entity representing a financial transaction in the UCO Bank system.
 */
@Entity
@Table(name = "transactions", indexes = {
    @Index(name = "idx_transaction_reference", columnList = "transactionReference", unique = true),
    @Index(name = "idx_from_account", columnList = "fromAccountId"),
    @Index(name = "idx_to_account", columnList = "toAccountId"),
    @Index(name = "idx_created_at", columnList = "createdAt")
})
public class Transaction {
    
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;
    
    @NotNull
    @Column(unique = true, nullable = false, length = 50)
    private String transactionReference;
    
    @NotNull
    @Column(nullable = false)
    private UUID fromAccountId;
    
    @NotNull
    @Column(nullable = false)
    private UUID toAccountId;
    
    @NotNull
    @Positive
    @Column(nullable = false, precision = 19, scale = 2)
    private BigDecimal amount;
    
    @NotNull
    @Column(nullable = false, length = 3)
    private String currency = "INR";
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private TransactionType transactionType;
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private TransactionStatus status = TransactionStatus.PENDING;
    
    @Column(length = 500)
    private String description;
    
    @Column(length = 500)
    private String remarks;
    
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @Column
    private LocalDateTime completedAt;
    
    @Column(length = 50)
    private String deviceId;
    
    @Column(length = 500)
    private String signature;
    
    @Column(length = 50)
    private String nonce;
    
    @Column
    private LocalDateTime signatureTimestamp;
    
    @Version
    private Long version;
    
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (transactionReference == null) {
            transactionReference = "TXN" + System.currentTimeMillis() + UUID.randomUUID().toString().substring(0, 8).toUpperCase();
        }
    }
    
    // Getters and Setters
    public UUID getId() {
        return id;
    }
    
    public void setId(UUID id) {
        this.id = id;
    }
    
    public String getTransactionReference() {
        return transactionReference;
    }
    
    public void setTransactionReference(String transactionReference) {
        this.transactionReference = transactionReference;
    }
    
    public UUID getFromAccountId() {
        return fromAccountId;
    }
    
    public void setFromAccountId(UUID fromAccountId) {
        this.fromAccountId = fromAccountId;
    }
    
    public UUID getToAccountId() {
        return toAccountId;
    }
    
    public void setToAccountId(UUID toAccountId) {
        this.toAccountId = toAccountId;
    }
    
    public BigDecimal getAmount() {
        return amount;
    }
    
    public void setAmount(BigDecimal amount) {
        this.amount = amount;
    }
    
    public String getCurrency() {
        return currency;
    }
    
    public void setCurrency(String currency) {
        this.currency = currency;
    }
    
    public TransactionType getTransactionType() {
        return transactionType;
    }
    
    public void setTransactionType(TransactionType transactionType) {
        this.transactionType = transactionType;
    }
    
    public TransactionStatus getStatus() {
        return status;
    }
    
    public void setStatus(TransactionStatus status) {
        this.status = status;
    }
    
    public String getDescription() {
        return description;
    }
    
    public void setDescription(String description) {
        this.description = description;
    }
    
    public String getRemarks() {
        return remarks;
    }
    
    public void setRemarks(String remarks) {
        this.remarks = remarks;
    }
    
    public LocalDateTime getCreatedAt() {
        return createdAt;
    }
    
    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }
    
    public LocalDateTime getCompletedAt() {
        return completedAt;
    }
    
    public void setCompletedAt(LocalDateTime completedAt) {
        this.completedAt = completedAt;
    }
    
    public String getDeviceId() {
        return deviceId;
    }
    
    public void setDeviceId(String deviceId) {
        this.deviceId = deviceId;
    }
    
    public String getSignature() {
        return signature;
    }
    
    public void setSignature(String signature) {
        this.signature = signature;
    }
    
    public String getNonce() {
        return nonce;
    }
    
    public void setNonce(String nonce) {
        this.nonce = nonce;
    }
    
    public LocalDateTime getSignatureTimestamp() {
        return signatureTimestamp;
    }
    
    public void setSignatureTimestamp(LocalDateTime signatureTimestamp) {
        this.signatureTimestamp = signatureTimestamp;
    }
    
    public Long getVersion() {
        return version;
    }
    
    public void setVersion(Long version) {
        this.version = version;
    }
}

File: sfe/app/src/main/java/com/aegis/sfe/data/model/Transaction.kt
package com.aegis.sfe.data.model

import java.math.BigDecimal
import java.time.LocalDateTime

data class Transaction(
    val id: String,
    val transactionReference: String,
    val fromAccountId: String?,
    val toAccountId: String?,
    val amount: BigDecimal,
    val currency: String,
    val transactionType: TransactionType,
    val status: TransactionStatus,
    val description: String?,
    val remarks: String?,
    val createdAt: LocalDateTime,
    val completedAt: LocalDateTime?,
    val deviceId: String?,
    val signature: String?,
    val nonce: String?
)

enum class TransactionType {
    TRANSFER,
    DEPOSIT,
    WITHDRAWAL,
    PAYMENT,
    REFUND
}

enum class TransactionStatus {
    PENDING,
    PROCESSING,
    COMPLETED,
    FAILED,
    CANCELLED,
    REFUNDED
}

data class TransferRequest(
    val fromAccount: String,
    val toAccount: String,
    val amount: BigDecimal,
    val currency: String = "INR",
    val description: String?,
    val remarks: String?
)

data class TransferResponse(
    val transactionReference: String,
    val status: String,
    val amount: BigDecimal,
    val currency: String,
    val fromAccount: String,
    val toAccount: String,
    val timestamp: String,
    val message: String
)

data class TransactionHistoryResponse(
    val content: List<TransactionResponse>,
    val pageable: PageableResponse,
    val totalElements: Long,
    val totalPages: Int
)

data class TransactionResponse(
    val id: String,
    val transactionReference: String,
    val fromAccountId: String?,
    val toAccountId: String?,
    val amount: BigDecimal,
    val currency: String,
    val transactionType: String,
    val status: String,
    val description: String?,
    val remarks: String?,
    val createdAt: String,
    val completedAt: String?,
    val deviceId: String?,
    val signature: String?,
    val nonce: String?
) {
    fun toTransaction(): Transaction {
        return Transaction(
            id = id,
            transactionReference = transactionReference,
            fromAccountId = fromAccountId,
            toAccountId = toAccountId,
            amount = amount,
            currency = currency,
            transactionType = TransactionType.valueOf(transactionType),
            status = TransactionStatus.valueOf(status),
            description = description,
            remarks = remarks,
            createdAt = LocalDateTime.parse(createdAt),
            completedAt = completedAt?.let { LocalDateTime.parse(it) },
            deviceId = deviceId,
            signature = signature,
            nonce = nonce
        )
    }
}

data class PageableResponse(
    val pageNumber: Int,
    val pageSize: Int
)